Tls cert race/core 9178/v25.1.x

michael-redpanda opened the pull request:
Fixes: CORE-9178

This addresses a possible race condition that some customers have witnessed where Seastar was not returning the correct certificate and this caused authorization failures on the broker. The fix is to use the correct OpenSSL API (SSL_get1_peer_certificate) rather than the cert held by the callback.

Note: We will backport this to v24.3 and v24.2 to address this bug in all OpenSSL implementations

michael-redpanda commented:
CI failures in two categories:

* CPU profiler unit test flakiness (I believe #174 is dealing with that)
* Failure in building rpc_test.cc - code doesn't touch this so it's probably something that's already there

StephanDollberg commented:
> Failure in building rpc_test.cc - code doesn't touch this so it's probably something that's already there

Should be fixed by https://github.com/redpanda-data/seastar/pull/176

michael-redpanda commented:
By the end of our three PRs we may get one green run ;)

michael-redpanda commented:
CI Failures:
* Modules fails to build
* cpu_profiler still flaky

BenPope commented on src/net/ossl.cc:
```diff
@@ -1105,25 +1105,21 @@ class session : public enable_shared_from_this<session>, public session_impl {
-                            case SSL_R_UNEXPECTED_EOF_WHILE_READING:
-                                // well in this situation, the remote end closed
-                                ERR_clear_error();
+                            auto error_codes = get_all_ossl_errors();
+                            if (contains_ossl_error(error_codes, ERR_LIB_SSL, SSL_R_UNEXPECTED_EOF_WHILE_READING)) {
```
Should this pattern be used in other cases where `ERR_peek_error() != 0`?

In all the other cases, `ERR_clear_error()` is also used.

michael-redpanda replied:
Yeah that's probably fair

pgellert commented on src/net/ossl.cc:
```diff
@@ -1105,25 +1105,21 @@ class session : public enable_shared_from_this<session>, public session_impl {
-                                ERR_clear_error();
+                            }
+
+                            if (contains_ossl_error(error_codes, ERR_LIB_SSL, SSL_R_PEER_DID_NOT_RETURN_A_CERTIFICATE) ||
+                                contains_ossl_error(error_codes, ERR_LIB_SSL, SSL_R_CERTIFICATE_VERIFY_FAILED) ||
+                                contains_ossl_error(error_codes, ERR_LIB_SSL, SSL_R_NO_CERTIFICATES_RETURNED)) {
```
2 questions:
* Could there be any cases when any of these errors are recoverable, ie. when they are not a problem even though they are present on the error stack? I assume no.
* Is `SSL_R_NO_CERTIFICATES_RETURNED` always an error or only when `verify()` throws?

michael-redpanda replied:
> Could there be any cases when any of these errors are recoverable, ie. when they are not a problem even though they are present on the error stack? I assume no.

No, they indicate a certificate verification error either due to the server sending a bad (or no) cert to the client or the client sent a bad (or no) cert to the server when VERIFY_PEER is set.

> Is SSL_R_NO_CERTIFICATES_RETURNED always an error or only when verify() throws?

This error only appears to occur on SSL3 (so <TLS1.0) connections, it's here for completeness.  It should be throwing an error in `verify()`.

pgellert commented on src/net/ossl.cc:
```diff
@@ -1636,12 +1644,24 @@ class session : public enable_shared_from_this<session>, public session_impl {
 
-    std::optional<session_dn> extract_dn_information() const {
-        const auto& peer_cert = get_peer_certificate();
+    using is_verification_error_t = bool_class<struct is_verification_error_tag>;
```
nit (happy to defer to when upstreaming / ignore): `is_verification_error_t` -> `is_verification_error`

pgellert commented on src/net/ossl.cc:
```diff
@@ -1329,7 +1337,7 @@ class session : public enable_shared_from_this<session>, public session_impl {
         if (res != X509_V_OK) {
             auto stat_str(X509_verify_cert_error_string(res));
-            auto dn = extract_dn_information();
+            auto dn = extract_dn_information(is_verification_error_t::yes);
```
Would it make sense to try to keep track of not just the last certificate but all the certificates that have been seen through `verify_callback` and include it in this verification error?

michael-redpanda replied:
That could be an unbounded list of certs.  We really only care about the last cert seen that caused the verification error.  It will step through the chain so if any in the chain (root ca, intermediate, or leaf) cause an error the verification will immediately fail.

michael-redpanda commented on tests/unit/tls_test.cc:
```diff
@@ -1046,8 +1046,12 @@ SEASTAR_THREAD_TEST_CASE(test_reload_certificates) {
+                // during the handshake before the client has fully
+                // closed its connection, then the get() call will
+                // succeed by return an empty buffer indicating EOF
+                BOOST_REQUIRE(res.size() == 0);
```
updated the wording

michael-redpanda commented:
Force push [7d7c0c8](https://github.com/redpanda-data/seastar/commit/7d7c0c8ec90af5b3c3f3bdfbc3f9af209eef49c3):

* Changed name of bool_class type
* Fixed wording in `tls_test.cc`
* Using new error check mechanism in place of `ERR_peek_error`

IoannisRP reviewed (approved):
lgtm

BenPope reviewed (approved):
LGTM
